onnx/backend/test/case/node/clip.py

```diff
@@ -16,70 +16,74 @@ class Clip(Base):
     def export():  # type: () -> None
         node = onnx.helper.make_node(
             'Clip',
-            inputs=['x'],
+            inputs=['x', 'min', 'max'],
             outputs=['y'],
-            min=-1.0,
-            max=1.0
         )
 
         x = np.array([-2, 0, 2]).astype(np.float32)
-        y = np.clip(x, -1, 1)  # expected output [-1., 0., 1.]
-        expect(node, inputs=[x], outputs=[y],
+        min_val = np.array(-1).astype(np.float32)
```

Comment:
this one should be a scaler, instead of an array. min_val = np.float32(-1)?

Reply:
I think we should support both scalar and 1-d tensor. That seems to be the consensus for the current onnx spec of the definition scalar. 

For example
https://github.com/onnx/onnx/blob/master/docs/Operators.md#size

> Size
Takes a tensor as input and outputs a int64 **scalar** that equals to the total number of elements of the input tensor.
...
```
node = onnx.helper.make_node(
    'Size',
    inputs=['x'],
    outputs=['y'],
)
x = np.array([
    [1, 2, 3],
    [4, 5, 6],
]).astype(np.float32)
y = np.array(6).astype(np.int64)
expect(node, inputs=[x], outputs=[y],
       name='test_size_example')
```

https://github.com/onnx/onnx/blob/master/docs/Operators.md#nonmaxsuppression

> score_threshold (optional) : tensor(float)
Float representing the threshold for deciding when to remove boxes based on score. It is a scalar
```
iou_threshold = np.array([0.5]).astype(np.float32)
score_threshold = np.array([0.0]).astype(np.float32)
selected_indices = np.array([[0, 0, 3], [0, 0, 0], [0, 0, 5]]).astype(np.int64)
```

Reply:
If you allow tensors, it means you need to check the shape in runtime because ONNX can have dynamic shapes. Whenever possiblre, I prefer strongly-typed way, so I'd suggest to follow Ke's comment. 

Previously, we didn't do a good job to emphasize the difference between scalar and 1-element tensor, but it doesn't mean we can't starting improving it from now on.

:)

Reply:
I agree that runtime will need to check the shape, but that is introduced by changing min/max from attributes to tensors. It has nothing to do with whether it is a scalar(0-rank tensor) or a 1-element tensor.

scalar vs 1-elem tensor is a bigger issue and should be discussed separately. It is not immediately clear to me which one is more preferable. That being said, I think we should use 'tensor' as description in the spec, avoid using the ambiguous 'scalar'.

Reply:
In ONNX, the definition of scalar is a tensor with empty shape. Since `min` and `max` attributes are scalars (they are float attribute instead of floats attribute), I don't feel we should use 1-element tensor to store them.

Reply:
I agree with @wschin and @linkerzhang : it is better to consistently use a tensor with an empty shape as a scalar, instead of a 1D tensor with 1 element. We have not been doing this consistently, but this will potentially create problems later on. In this case, since we are specifying a unit-test-case, we should use zero-dimensional tensors to represent scalars in the test-case. Otherwise, implementations will start supporting both zero-dimensional and 1D tensors (to get past the test-case) and then when somebody wants to do code-generation they will be unable to represent the value using a primitive value in the generated-code … they will be forced to support a boxed representation, making things inefficient.

Reply:
Note that people can always use a Reshape to coerce any tensor with 1 element into a scalar. The idea is that people should pay the cost of this kind of representation-conversion only when needed. Strong typing will enable zero-cost when typing guarantees the representation is already the correct one.

Reply:
As long as the test-cases here represent the corresponding scalars with empty shape (is this correct?), we should be fine here.

Reply:
Thanks for the review @wschin @gramalingam @linkerzhang ! I agree we should have a clear and strict definition of scalar.